csharp/ql/src/semmle/code/csharp/dataflow/internal/rangeanalysis/SignAnalysisSpecific.qll

```diff
@@ -0,0 +1,278 @@
+    // Keep them in sync.
+    not e instanceof AssignExpr and
+    not e instanceof AssignOperation and
+    not e instanceof UnaryPlusExpr and
+    not e instanceof PostIncrExpr and
+    not e instanceof PostDecrExpr and
+    not e instanceof PreIncrExpr and
+    not e instanceof PreDecrExpr and
+    not e instanceof UnaryMinusExpr and
+    not e instanceof ComplementExpr and
+    not e instanceof AddExpr and
+    not e instanceof SubExpr and
+    not e instanceof MulExpr and
+    not e instanceof DivExpr and
+    not e instanceof RemExpr and
+    not e instanceof BitwiseAndExpr and
+    not e instanceof BitwiseOrExpr and
+    not e instanceof BitwiseXorExpr and
+    not e instanceof LShiftExpr and
+    not e instanceof RShiftExpr and
+    not e instanceof ConditionalExpr and
+    not e instanceof RefExpr and
+    not e instanceof LocalVariableDeclAndInitExpr and
+    not e instanceof SwitchCaseExpr and
+    not e instanceof CastExpr and
+    not e instanceof SwitchExpr and
+    not e instanceof NullCoalescingExpr
+  }
+
+  Sign specificSubExprSign(Expr e) {
+    // The expression types that are handled here should be excluded in `unknownIntegerAccess`.
+    // Keep them in sync.
+    result = exprSign(e.(AssignExpr).getRValue())
+    or
+    result = exprSign(e.(AssignOperation).getExpandedAssignment())
+    or
+    result = exprSign(e.(UnaryPlusExpr).getOperand())
+    or
+    result = exprSign(e.(PostIncrExpr).getOperand())
+    or
+    result = exprSign(e.(PostDecrExpr).getOperand())
+    or
+    result = exprSign(e.(PreIncrExpr).getOperand()).inc()
+    or
+    result = exprSign(e.(PreDecrExpr).getOperand()).dec()
+    or
+    result = exprSign(e.(UnaryMinusExpr).getOperand()).neg()
+    or
+    result = exprSign(e.(ComplementExpr).getOperand()).bitnot()
+    or
+    e =
+      any(DivExpr div |
+        result = exprSign(div.getLeftOperand()) and
+        result != TZero() and
+        div.getRightOperand().(RealLiteral).getValue().toFloat() = 0
+      )
+    or
+    exists(Sign s1, Sign s2 | binaryOpSigns(e, s1, s2) |
+      e instanceof AddExpr and result = s1.add(s2)
+      or
+      e instanceof SubExpr and result = s1.add(s2.neg())
+      or
+      e instanceof MulExpr and result = s1.mul(s2)
+      or
+      e instanceof DivExpr and result = s1.div(s2)
+      or
+      e instanceof RemExpr and result = s1.rem(s2)
+      or
+      e instanceof BitwiseAndExpr and result = s1.bitand(s2)
+      or
+      e instanceof BitwiseOrExpr and result = s1.bitor(s2)
+      or
+      e instanceof BitwiseXorExpr and result = s1.bitxor(s2)
+      or
+      e instanceof LShiftExpr and result = s1.lshift(s2)
+      or
+      e instanceof RShiftExpr and result = s1.rshift(s2)
+    )
+    or
+    result = exprSign(e.(ConditionalExpr).getAChild())
+    or
+    result = exprSign(e.(NullCoalescingExpr).getAChild())
+    or
+    result = exprSign(e.(SwitchExpr).getACase().getBody())
+    or
+    result = exprSign(e.(CastExpr).getExpr())
```

Comment:
This can be improved a bit
```ql
    e =
      any(CastExpr ce |
        result = exprSign(ce.getExpr()) and
        if ce.getTargetType() instanceof UnsignedIntegralType then result != TNeg() else any()
      )
```

Reply:
(But feel free to do this in the follow-up PR that handles unsigned types.)